### PR TITLE
Add transformer for JSON strings

### PR DESCRIPTION
**Pull request by tiurin**
### Motivation

Some services response contains JSON strings within JSON strings, e.g. Cloudwatch Logs for EventBridge Pipes events.

We do have [first level JSON strings parsed](https://github.com/localstack/localstack-snapshot/blob/302c9937a16a607bf93b24aaa50b069f0ec0d9e6/localstack_snapshot/snapshots/prototype.py#L274) but we don't go recursively into parsed JSONs. This leads to the need of extracting and parsing such payloads separately in the test because otherwise only regex transformers can be applied to such string. Ideally, snapshot library should take care of this. See corresponding [discussion](https://github.com/localstack/localstack-ext/pull/3956#discussion_r1928467194) (thanks @gregfurman for inspiration!).

### Changes

Starting to parse nested JSONs in common transformation flow is a breaking change for existing recorded snapshots. This PR proposes non-breaking approach by introducing a separate transformer that can be applied on demand. This draft shows a usage example: https://github.com/localstack/localstack-ext/pull/3971

### CODE REVIEW

File: tests/test_transformer.py [hunk 1]
```diff
@@ -236,6 +237,54 @@ def test_text(self, value):
             output = sr(output)
         assert json.loads(output) == expected
 
+    @pytest.mark.parametrize(
+        "input_value,transformed_value",
+        [
+            pytest.param('{"a": "b"}', {"a": "b"}, id="simple_json_object"),
```
Comment (tiurin) [inline tests/test_transformer.py:243]: TODO: add test case for empty value

File: localstack_snapshot/snapshots/transformer.py [hunk 2]
```diff
@@ -360,3 +362,65 @@ def replace_val(s):
...
+        elif isinstance(input_data, list):
+            return self._transform_list(input_data, ctx=ctx)
+        return input_data
+
+    def _transform_dict(self, input_data: dict, ctx: TransformContext = None) -> dict:
+        for k, v in input_data.items():
+            if k == self.key and isinstance(v, str):
+                try:
+                    SNAPSHOT_LOGGER.debug(f"Replacing string value of {k} with parsed JSON")
+                    json_value = json.loads(v)
```
Comment (joe4dev) [inline localstack_snapshot/snapshots/transformer.py:393]: as discussed, we're missing a test and fix for empty strings here `""` to avoid verbose exception logging

File: localstack_snapshot/snapshots/transformer.py [hunk 3]
```diff
@@ -360,3 +362,65 @@ def replace_val(s):
             f"Registering text pattern '{self.text}' in snapshot with '{self.replacement}'"
         )
         return input_data
+
+
+class JsonStringTransformer:
+    """
+    Parses JSON string at key.
+    Additionally, attempts to parse any JSON strings inside the parsed JSON
```
[the remark this answers is not in the document]
Comment (joe4dev) [reply localstack_snapshot/snapshots/transformer.py:370]: In particular, highlighting the difference to the default behavior might be most relevant, clarifying that it adds extra coverage for JSON lists starting with `[` such as `[{},{}]`

File: localstack_snapshot/snapshots/transformer.py [hunk 4]
```diff
@@ -360,3 +362,65 @@ def replace_val(s):
...
+    """
+
+    key: str
+
+    def __init__(self, key: str):
```
Comment (joe4dev) [inline localstack_snapshot/snapshots/transformer.py:375]: Idea: To facilitate an easy transition, it could be nice to have an easy way to apply this transformer globally, for example, when not providing a key.


sidenote: it ultimately comes down to separating anchors (e.g., JSON path selectors) from actual transformer functionality, but that's far out of scope :)

**Review by joe4dev — COMMENTED** [review]
Thank you for thinking about a backward-compatible way of adding such a neat feature, making lists of JSON-encoded snapshots much easier to handle 👏 .

As discussed, the main open part is handling the special edge case of empty strings. Otherwise, I added some suggestions and ideas for future iterations 📈 



Idea 💡: One of the disadvantages of automatic parsing is that we are loosing structural parity information, we can:
* a) highlight this in our snapshot testing guidelines
* b) implement a technical solution at some point such as using some kind of metadata to store the raw value or structural information (e.g., `key__META__`: {"json": True})



Sidenotes (as discussed and partly related):
* Clarify dev installation instructions with troubleshooting for the weird editable install issue (separate quick docs PR)
* Improve Utility inheritance structure such that LocalStack Core only adds AWS-specific transformer utils (e.g., lambda_api) and utils such as `sorting` are implemented in the snapshot library

**Comment by joe4dev** [comment]
👍 I tested this change against the localstack-ext PR https://github.com/localstack/localstack-ext/pull/3971 with the following changes (not necessary anymore once the Utility re-use hierarchy is fixed) and can confirm it works as expected (apart from the minor empty string stack trace):

```diff
         snapshot.add_transformer(
-            snapshot.transform.sorting(
+            SortingTransformer(
                 "events",
                 lambda event: (event["message"]["executionId"], event["message"]["messageType"]),
             )
         )
-        snapshot.add_transformer(snapshot.transform.json_string("payload"))
+        # snapshot.add_transformer(snapshot.transform.json_string("payload"))
+        snapshot.add_transformer(JsonStringTransformer("payload"))
```


✅ There is no regression (identity check) using this transformer via `snapshot.add_transformer(JsonStringTransformer("Document"))` in the X-Ray test `tests.aws.services.lambda_.test_lambda_xray.TestLambdaXrayIntegration.test_basic_xray_integration` (X-Ray commonly uses nested JSON documents, but that's already handled by the default transformer)

File: localstack_snapshot/snapshots/transformer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tiurin) [reply localstack_snapshot/snapshots/transformer.py:385]: > docs: Are there any limitations (e.g., we're losing parity information that the original response is actually JSON-encoded and should have some tests covering that)?

I don't think we're losing parity information here, at least not any more than with other transformers. Conceptually, it is the same as with any other transformer - we apply some transformation to data that can't be compared between AWS and Localstack test runs, or between different runs.
Applying this transformer is an explicit choice of its user. This means that the user decided to parse JSON string at specific key in order to reliably compare its contents - after decoding other transformers like timestamp, UUID, ARNs, etc. will automatically apply.

[the remark this answers is not in the document]
Comment (tiurin) [reply localstack_snapshot/snapshots/transformer.py:385]: Updated docstring as suggested, thanks!

File: localstack_snapshot/snapshots/transformer.py [hunk 5]
```diff
@@ -360,3 +362,65 @@ def replace_val(s):
...
+            input_data = [self._transform_nested(item) for item in input_data]
+        if isinstance(input_data, dict):
+            for k, v in input_data.items():
+                input_data[k] = self._transform_nested(v)
+        if isinstance(input_data, str) and input_data.strip().startswith(("{", "[")):
+            try:
+                json_value = json.loads(input_data)
+                input_data = self._transform_nested(json_value)
+            except JSONDecodeError:
+                SNAPSHOT_LOGGER.debug(
```
[the remark this answers is not in the document]
Comment (tiurin) [reply localstack_snapshot/snapshots/transformer.py:451]: I'd err on logging side. By explicitly applying this transformer user expects correct nested JSON parsing, unlike with top-level behavior that is always a best effort, and an implicit one. If there is anything wrong with JSON parsing here I'd rather let user know about it.

File: localstack_snapshot/snapshots/transformer.py (hunk 2 above)
Comment (tiurin) [reply localstack_snapshot/snapshots/transformer.py:421]: Good point, thanks! Added several empty values test cases - empty string, empty list, empty object. Also added extra check for string to condition.

File: localstack_snapshot/snapshots/transformer.py (hunk 4 above)
Comment (tiurin) [reply localstack_snapshot/snapshots/transformer.py:403]: This transformer handles a specific case and I'd not recommend to apply it globally.

File: localstack_snapshot/snapshots/transformer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (joe4dev) [reply localstack_snapshot/snapshots/transformer.py:385]: Great clarifications and comparison in the docstring 💯 👍 
Our future selfes will thank us :)

File: localstack_snapshot/snapshots/transformer.py (hunk 4 above)
Comment (joe4dev) [reply localstack_snapshot/snapshots/transformer.py:403]: Good point. Let's not introduce unnecessary complexity 👍

File: localstack_snapshot/snapshots/transformer.py (hunk 2 above)
Comment (joe4dev) [reply localstack_snapshot/snapshots/transformer.py:421]: Thanks for adding extra coverage for these cases 👍

**Review by joe4dev — APPROVED** [review]
Thank you for pushing this over the line and extending unit test coverage @tiurin 👍 

I re-tested the following two cases:

🟢 There is no regression (identity check) using this transformer via `snapshot.add_transformer(JsonStringTransformer("Document"))` in the X-Ray test `tests.aws.services.lambda_.test_lambda_xray.TestLambdaXrayIntegration.test_basic_xray_integration` (X-Ray commonly uses nested JSON documents, but that's already handled by the default transformer)

🟡/🟢 The localstack-ext PR https://github.com/localstack/localstack-ext/pull/3971 requires the following changes. This is primarily syntactic sugar for importing. The behavior works properly for the payload field 👍 :

    ```diff
             snapshot.add_transformer(
    -            snapshot.transform.sorting(
    +            SortingTransformer(
                     "events",
                     lambda event: (event["message"]["executionId"], event["message"]["messageType"]),
                 )
             )
    -        snapshot.add_transformer(snapshot.transform.json_string("payload"))
    +        # snapshot.add_transformer(snapshot.transform.json_string("payload"))
    +        snapshot.add_transformer(JsonStringTransformer("payload"))
    ```

🚀  This unblocks https://github.com/localstack/localstack-ext/pull/3971

@tiurin Are you planning to adjust the utility re-use hierarchy in a follow-up, or do we advocate for direct imports?

**Comment by tiurin** [comment]
> Are you planning to adjust the utility re-use hierarchy in a follow-up, or do we advocate for direct imports?

@joe4dev I was about to propose advocating for direct imports but then [`key_value`](https://github.com/localstack/localstack/blob/727155eaa76367684a7b1878ef5c58275dbaa3a3/localstack-core/localstack/testing/snapshots/transformer_utility.py#L67) factory actually does some additional processing for transformer initialization. Also there is a [`resource_name`](https://github.com/localstack/localstack/blob/727155eaa76367684a7b1878ef5c58275dbaa3a3/localstack-core/localstack/testing/snapshots/transformer_utility.py#L73) one. Let's analyze and elaborate separately.

One thing I surely don't like is having `snapshot.add_transformer(snapshot.transform.` - passing the parameter from snapshot session to a function also defined in a snapshot session. It's quite difficult to read every time I see it. So definitely some utility and fixture rework is welcome, let's keep this conversation going outside this PR.

In any case, as long as we have the utility, I think [`TransformerUtility`](https://github.com/localstack/localstack/blob/727155eaa76367684a7b1878ef5c58275dbaa3a3/localstack-core/localstack/testing/snapshots/transformer_utility.py#L49) in **localstack codebase** should inherit the one **snapshot library**. Right now both are imported in different places across community and pro repos and having the same name they are so easy to confuse. [Snapshot fixture](https://github.com/localstack/localstack/blob/fd727d533f66392fcc9ac7417657fe02e5a9e0e1/tests/aws/conftest.py#L113) even issues a warning about type mismatch when I open it in PyCharm. Also, `TransformerUtilityExt` inherits the utility from localstack, so right now it doesn't get new methods added to the library one. That's why example PR in ext doesn't compile - this is how I identified the issue first, just didn't change the example yet. 
I will definitely address this issue in upcoming PRs, it is a bit of a time bomb.
